### ISSUE TITLE
DRAFT: [RFC] Using Terrareg as a registry for OpenTofu registry

### ISSUE THREAD

**Issue by MatthewJohn** [opentofu/opentofu#791]
### Summary

(apologies, I was unaware of the "RFC" issue template before starting the compilation of this and will adapt)

This issue is to discuss how Terrareg could be used as a registry for OpenTofu, as per the requirements in #258

Rather than the "approach" driven RFCs, by others, this is more geared towards the "off the shelf" suggestions that have been made and how Terrareg could fit into the OpenTofu eco-system.

I aim to outline how it _could_ be used, purely to demonstrate the fact, rather than promote the use of it.

Terrareg is a pre-built solution, regarding the core application, though I will discuss what could be implemented for deployment in a larger scale environment.

I see the suggestions of implementations for the registry going very differently to how Hashicorp runs (on purpose of course):
 * Hashicorp runs a (relatively) owership-based solution:
   * A user owns a Github organisation, they add their providers (or modules) and the registry automatically indexes these
 * OpenTofu registry (in some manor) replicates providers and modules that have already been published to Hashicorp's registry (or via some other means), but the ownership of these is taken, for the most part, out of the control of the Github owner's hands (to facility providers/modules) to users that want to migrate before the providers migrate.

Terareg natively compensates for the former of these. Though, there is no real technical limiation as to why it can't support both methodologies. The advantage of supporting both is that, once providers _do_ decide to migrate, the functionality for them to maintain their owner resources is already there. 

### Problem Statement

TBC

### User-facing description

TBC

## Outline of Design

Terrareg is a python application, built in standard monolith fashion and can run in ECS/fargate.

It uses a mysql database, which can use RDS.

For python modules, it uses Flask as a web server, SQLAlchemy for database interactions and various open source tools for extracting information about providers/modules.

### Drop-in replacement

```
It must be a drop-in replacement for resolving providers and modules.
That is existing versions of these have to be available; new versions have to be picked up without action on the provider/module author’s side.
```

Terrareg currently has two models of working with users:
 * Admin/OIDC authentication, which provides mappings to pre-defined groups. In term, these provide access given "namespaces". This would allow internal members of OpenTofu team the ability to authenticate and create namespaces and modules on users' behalf
 * Github authentication, which automatically creates/provides access to any organisations that the user is a member of (and for their own user profile)

Neither of these would provide an automated solution for importing namespaces, modules and providers automatically.
Though, creating a batch job that iterates through a list of known Github orgs/repositories wouldn't be too difficult.

This could also be alterated to query the hashicorp APIs for a list of providers/modules to be imported.

Importing of new module/provider versions also makes use of hook APIs, which are normally present in the repositories - though this is not required, and a similar batch job could be created to query for any newly published versions.

TBC: How would this authentication work and what affect would it have on rate-limiting

### GPG Key Security

```
It must make available a way for authors to submit/revoke/update their public keys, in a way compatible with existing provider signatures.

Handling of public keys and signatures must provide security gains equal to or above those of the HashiCorp registry. 
 *  To be more specific, in the case of the HashiCorp registry, capturing the session of an organization GitHub admin would let you add a new key and publish malicious artifacts through GitHub actions.
```

Natively, a user is able to authenticate via Github, which will provide admin access to any namespaces, based on any organisations that they are admins over.
From here, they are able to upload/revoke any GPG keys.

From a non-maintained organisation point-of-view (one that has been created by OpenTofu for replication purposes), a batch job could be written to query Hashicorp Terraform APIs to replicate GPGs, relatively easily.

### Artifact hosting

```
Even though we can start by hosting artifacts (providers) in GitHub releases, the solution should allow swapping out the artifact download URL to a 3rd party on a provider-by-provider basis in a provider-author-controlled way (as in, it shouldn’t require the end user to change config on their side). 

    E.g. if we see the aws provider is downloaded a lot, and GitHub release artifacts for whatever reason cannot be used anymore for it, we need to be able to switch to a CDN or other download location, without any action required from the CLI user.
```

For providing artifacts for providers/modules, Terrareg can be used in two ways:
 * Provide a redirect URL
 * Serve the artifact

For modules:
 * Terrareg can redirect terraform download the repository; or
 * Serve an archive artifact of the module, generated during indexing

For providers:
 * Terrareg can redirect terraform to the release artifacts in github for each resource; or
 * Serve an archive artifact of each of the release artifacts, stored during indexing (in progress)

Whilst using the Github repositories/release artifacts would provide the least "load" on the registry, cloudfront could be placed in-front of the artifact endpoints, which would alleviate load on Github (though cloudfront costs would be questionable)

### Namespace redirects

```
Needs to redirect hashicorp-namespaced providers to the opentofu namespace.
```

Namespace and module name redirection is supported by Terrareg. When a namespace is renamed, it redirects a user to the appropriate URL (from the UI) and the API accepts requests to the namespaces that have not yet been replaced.
A namespace could be setup for hashicorp and renamed to opentofu, which would add this redirect.
However, this would _not_ stop an owner of the hashicorp organisation from authenticating, which would remove the redirect (@TODO This restriction would be implemented)

### Provider version warnings

```
Must allow for warnings to be attached to provider version metadata. E.g. the registry currently serves a custom warning when a user tries to fetch the long-deprecated terraform provider. See this issue for more details.
```

Warnings are available via the provider model.
However, these are not currently modifiable via any APIs/UI, though this could be implemented. Though, they can be modified via the database.

TBC: These are only implemented in the v2 APIs - need to verify where these are defined in the v1 APIs

### Rate limiting and identity

```
It must support a single “identity” running thousands of concurrent executions of Tofu (that list and fetch providers and modules) without getting rate-limited. 
    An identity could be an IP address, a logged-in user (if a login-gate is involved), a company, etc.
```

Terareg would make requests to Github in these situations:
 * Module version indexing - currently this is performed by anonymous Git clones
 * Provider version indexing - this currently uses user access key in the draft implement, but will be switched to app installation authentication. TBC: How will this work for providers that are added without organisation ownership?
 * Github authentication

TBC: Add information about API limits

### UI

```
Should provide provider and module discovery and documentation via a human-readable web UI, similar to the legacy Terraform registry (this must be part of the design, but can be implemented after the stable release). 

    Worth noting that it seems like the legacy Terraform registry hosts json files with metadata to provide required information for each provider/module, and then renders it client-side based on the json file for the currently-viewed provider/module.
```

The majority of Terrareg's unique feature is the information it attempts to provide via the user interface.
It provides a user interface:
 * Creating namespaces, providers and modules
 * Searching for modules (searching for providers is in-progress)
 * View details for modules, submodules, examples
 * View details for providers and documentation

For gathering information on module, it uses:
 * terraform CLI
 * terraform-docs
 * infracost
 * tfsec

For gathering information on providers, it uses:
 * tfplugindocs

### High availability

```
We’ll be expected to provide high availability (multi-region failover should be viable) and high security guarantees.
```

I would expect that Terrareg would primary be deployed to a single region, with cross-region replication.
ECS can be deployed with a single region and scaled. Note: Terrareg currently performs provider/module "extraction" within the web server, but this can (and should) be moved to an indepent worker pool with queue in-between.

Cloudfront can be used for caching most APIs (depending on the mode of operation (such as disabling "anonymous read access"), but assuming the authentication use-case of OpenTodu, this would be the case).
The exception to this cache would be the download APIs for providers/modules, as these play a role in recording provider/module download statistics.

However, if cloudfront was used in front of most APIs, except these and load was _still_ an issue, a lambda could be produced to provide the cached content, which would then added to queue to record the download statistics asyncronously.

### Downtime

```
We’d like to minimize the amount of maintenance work required to keep the registry up.
```
Terrareg attempts to make database modifications backwards compatible.
Though it is not currently performed, we can alter release cycles to ensure releases are backwards compatible, from a UI/API point of view, allowing releases to be gradually released to an ECS cluster (or similar).

Database changes are generally backwards compatible. When new features are added, they do not generally per data migrations, meaning that the application expects newly created columns to be empty without pre-population via a migration script (that are expected to be populated by the application).

### Open source

```
Implementation of the registry must be fully open-source.
```

It is - it is licensed under GNU General Public License v3.0 :)

### Download metrics

```
Download stats for each provider/module. 
    This is very useful for discoverability purposes.
```

Provider and module downloads are captured by Terrareg and are displayed on the homepage and for each provider/module, similar to Hashicorp

### v1 protocols

```
Implements the V1 registry protocol for providers and modules. 
    This means it will work with legacy Terraform and will not require CLI-side changes in OpenTofu.
```

The v1 registry protocols are implemented and Terraform (not tested with OpenTofu) works compltely.

### Mirror hosting

```
It should be easy to host mirrors of this registry.
```

Depending on the exact meaning of this requirement, the application can be scaled "horitzontally".
However, mirroring outside of the registry, extraction tools would be needed to dump information about each of the provider, modules and each of their version to be used in a mirrored application.

### v2 APIs

```
Implement the unofficial v2 API of the HashiCorp Terraform Registry 
```

Work has begun on implementing these v2 APIs for providers, but is in the very early stages.

### Rationale and alternatives

In my opinion (purely), Terrareg really tries to achive the UI-replacement for hashicorp.

### Downsides

Terrareg is a monolith application, meaning it comes with the downsides of this:
 * Only the core application is scalable
 * It heavily relies on a single database, though this can be scaled in RDS
 * It does not use any cloud-specific features out of the box (somewhat by-design), though adding optional support for these would not be "out of the question".
 * It is written in Python (you may draw your own conclusions from this)

Some pre-requisites for it's use include:
 * Complete work on implementing Provider support (see https://gitlab.dockstudios.co.uk/pub/terrareg/-/merge_requests/360 for completed/remaining tasks)
 * Moving indexing of providers/modules to a queue-based system and an external processing application.

Terrareg is largly written around an ownership model, where an owner of something (Github org etc.) will be maintaining their own resources (and can use hooks to automate this). From a replication point-of-view, automating this will largely need to be written into jobs that monitor and mimic the webhook behavoirs, as well as providing the automatic creation of providers/modules and namespaces, as they are created/wished to be mirrored. 

Code base qualtiy: There is plenty of the code base that could do with tidying up, but test coverage is aimed to be relatively high (and relatively thorough)

Terrareg may not be the best suited tool for the job - but: it can be adapted and, I think, it certainly has some selling points and (as unlikely as it is), I feel I'd be doing it a disservice by not at least submitting it.

### Proof of Concept

* A proof of concept is a available at https://registry-demo.mattsbit.co.uk/, though this does not currently contain any of the "provider" functionality, which has been in progress. This will be updated during the Weekend commencing 28th Oct.

**Comment by cube2222** [comment]
Hey @MatthewJohn! Let me start by saying - thanks a lot for creating this RFC and putting in the work to write this up. It's very appreciated!

Now, to my feedback, though I'd like to emphasize that this is my personal opinion - I'm not making the decision, the steering committee is. Overall, I do not personally think this is the right solution for the OpenTofu official registry, at least when compared to the others.

Terrareg really seems to be solving a very different use case and satisfying a different set of tradeoffs (don't get me wrong, it's a very notable and sensible use case and set of tradeoffs!) than what we're looking for. It really looks like we're trying to make Terrareg be something it's not here and fit it into this hole, instead of using it because it's "the right solution for the problem".

Some specifics:

### High-Availability

> I would expect that Terrareg would primary be deployed to a single region, with cross-region replication.

If I understand correctly, you mean that High-Availability would be accomplished by using multi-region SQL database replication. I understand that this is **doable** but multi-region Postgres replication definitely isn't a fun experience, nor does it help to minimize maintenance work and potential for outages.

> The exception to this cache would be the download APIs for providers/modules, as these play a role in recording provider/module download statistics.

This (and the version listing) are the main parts of the API that need to be very highly available, so are the endpoints we definitely need to cache. The remaining endpoints (docs, search, metrics) don't matter that much as far as HA goes, they are not critical for hot-path usage of OpenTofu.

This does mean that the approach to collecting download metrics simply based on requests served won't work.

### Drop-in replacement

It doesn't seem like the use-case of importing providers, modules, and gpg keys is supported yet, same for the use case of indexing those 3rd party projects and managing GitHub rate limits while doing so. It also seems like with the proposed solution, the work would be fully in the hands of the admins, and users wouldn't be able to submit a 3rd party provider for being indexed.

> Importing of new module/provider versions also makes use of hook APIs, which are normally present in the repositories - though this is not required, and a similar batch job could be created to query for any newly published versions.
> TBC: How would this authentication work and what affect would it have on rate-limiting

This is one of the main problems we are trying to solve here, so should in a sense be the starting point for an RFC, while here it's left as a TBC. I understand we can adapt terrareg to support this, I don't believe we'd be saving time by doing that.

### Ease of Maintenance

The amount of operational complexity here overall is **relatively** large, even for a single-region deployment - we need a SQL database, for which we need to manage backups, local replication, updates, access, etc.

It's worth noting that "relatively" here means relative to the Homebrew-like approach by @RLRabinowitz; I actually believe my own RFC - which uses a database, but not in the hot path - is too much operational complexity as well and I don't think we should use it.

### Mirroring

The point here was for 3rd parties (including companies for internal use) to easily be able to spin up a copy of the registry that has all the same contents.

### v2 APIs

That was actually removed from the list of requirements.

## Summary

The most important point to me is that Terrareg contains a lot of functionality and really is oriented around being a registry for user-owned provider and module hosting.

This would make sense if we were starting an ecosystem from scratch, but we're not. What we're trying to achieve, really (as you've said yourself in the RFC), is a community-maintained registry index of providers/modules maintained outside of it.

I'm sure we can adapt Terrareg to this use-case through a bunch of work, but the amount of this work won't be lower than writing just the required bits from scratch. Moreover, there'd be a huge amount of additional complexity that we don't need here that will be more potential for bugs and security vulnerabilities. Additionally, the work will have to be done in Python (while the core team's expertise is generally Go, since Go is what OpenTofu is written in).

I think your own words capture this well:
> Terrareg is largly written around an ownership model, where an owner of something (Github org etc.) will be maintaining their own resources (and can use hooks to automate this). From a replication point-of-view, automating this will largely need to be written into jobs that monitor and mimic the webhook behavoirs, as well as providing the automatic creation of providers/modules and namespaces, as they are created/wished to be mirrored.

We'd be trying to bend terrareg into something it wasn't made for.

Now again, to wrap this up, I believe terrareg is a great project, I believe it solves great use cases, I really appreciate you submiting this RFC and putting in the required work, I just don't believe it's the use case and set of tradeoffs we have and need here, and I don't think trying to adapt it to the use case will be less work than writing what we actually need from scratch.

**Comment by MatthewJohn** [comment]
Hey @cube2222,

Thank you for taking the time to respond - I appreciate it wasn't a great suggestion, though thought it was worth documenting and (since it appeared to be suggested by a couple of people in the original ticket) stating why it might not be a good candidate :) So thank you

> Mirroring
Thank you for explaining this this - it does make sense - on a side note, this is something I have on the roadmap as I think "it's good to have a method of replicating providers from one registry to another" :)

I do agree with your summary - on a slightly different angle, I am trying to achieve quite a few different use cases with Terrareg (and I know some people are using it in quite different ways), but as you say, for a large system, this can certainly have a deficit of being a prime way for bugs to be introduced.

> Besides
> ...
> Not sure what you / others think.

Thank you :) I do appreciate that and, personally, would love that :) :)

Thank you again for your time

Matt

**Comment by cube2222** [comment]
Alright, the steering committee has met yesterday, and as described in today's (2023-11-03) update in https://github.com/opentofu/opentofu/issues/258, they picked the Homebrew-like design.

I'll be closing this issue, but will create a separate issue for the documentation proposal.

Thanks once more for submitting the RFC @MatthewJohn!
